crates/red_knot_python_semantic/docs/kind_of_types.md

```diff
@@ -0,0 +1,254 @@
+# Kinds of types
+
+This document provides definitions for various kinds of types,
+with examples of Python types that satisfy these definitions.
+
+It doesn't attempt to be exhaustive.
+Only definitions that are useful for red-knot developers should be added to this document.
+
+## Singleton types
+
+A singleton type is a type for which it is known that there is
+(and can only ever be) exactly one known inhabitant of the type at runtime.
+For any singleton type in Python, the type's sole inhabitant will always exist
+at the same memory address for the entire duration of a Python program.
+
+Examples of singleton types in red-knot's model of Python include:
+
+- `types.EllipsisType` (the sole inhabitant is `builtins.Ellipsis`;
+    the constructor always returns the same object):
+
+    ```pycon
+    >>> from types import EllipsisType
+    >>> EllipsisType() is EllipsisType() is ... is Ellipsis
+    True
+    >>> id(EllipsisType()) == id(...)
+    True
+    ```
+
+- `types.NotImplementedType`
+    (in the same way as `types.EllipsisType` and `builtins.Ellipsis`,
+    the sole inhabitant is `builtins.NotImplemented`).
+
+- `None` (which can also be spelled as `Literal[None]`).
+    The sole inhabitant of the type is the runtime constant `None` itself.
+
+- `Literal[True]`: the sole inhabitant of the type is the constant `True`.
+
+- `Literal[False]`: the sole inhabitant of the type is the constant `False`.
+
+- `Literal[E.A]`, `Literal[E.B]` or `Literal[E.C]` for the following enum `E`
+    (the sole inhabitant of `Literal[E.A]` is the enum member `E.A`):
+
+    ```py
+    from enum import Enum, auto
+
+    class E(Enum):
+        A = auto()
+        B = auto()
+        C = auto()
+    ```
+
+- A "literal class type": a type representing a single known class (and none of its possible subclasses).
+
+- A "literal function type": a type that represents a single known function
+    (and excludes all other functions, even if they have the same signature).
+
+- A "literal module type": a type representing a single known module.
+
+Since it is known that all inhabitants of a given singleton type share the same memory address,
+singleton types in unions can be safely narrowed by identity,
+using [the operators `is` and `is not`](https://snarky.ca/unravelling-is-and-is-not/):
+
+```py
+def f(x: str | None):
+    if x is None:
+        ...  # x can be narrowed to `str`
+    else:
+        ...  # x can be narrowed to `None`
+```
+
+All singleton types are also sealed types, and closed types; nearly all are single-value types.
+(See below for definitions of these other concepts.)
+
+The number of singleton types in Python is theoretically infinite
+(for any given theoretical runtime object `x`, you could theorise a singleton type `Y` for which
+the sole inhabitant of `Y` would be the runtime object `x`). However, only a small fraction of
+the set of possible singleton types are recognised as being singleton types by red-knot's model.
```

Comment:
This phrasing feels slightly off to me. It's not about whether red-knot "recognizes them as singleton types" -- it's not as if red-knot recognizes them as some other kind of type instead! It's just that they are types for which red-knot has no internal representation (nor is there any way to spell them in a type expression.)